File: src/openfl/text/TextFormat.nim
import ./TextFormatAlign

type
  TextFormat* = ref object
    align*: TextFormatAlign
    blockIndent*: int
    boldInternal: bool
    bullet*: bool
    color*: int
    fontInternal: string
    indent*: int
    italicInternal: bool
    kerning*: bool
    leading*: int
    leftMargin*: int
    letterSpacing*: float64
    rightMargin*: int
    sizeInternal: int
    strikethrough*: bool
    tabStops*: seq[int]
    target*: string
    underline*: bool
    url*: string
    cacheKey: string

proc toCacheKey(self: TextFormat): string =
  self.cacheKey = self.fontInternal & $self.sizeInternal & $self.boldInternal & $self.italicInternal
  return self.cacheKey

proc font*(self: TextFormat): string = self.fontInternal
proc `font=`*(self: TextFormat, value: string) =
  if self.fontInternal != value:
    self.fontInternal = value
    discard self.toCacheKey()

proc size*(self: TextFormat): int = self.sizeInternal
proc `size=`*(self: TextFormat, value: int) =
  if self.sizeInternal != value:
    self.sizeInternal = value
    discard self.toCacheKey()

proc bold*(self: TextFormat): bool = self.boldInternal
proc `bold=`*(self: TextFormat, value: bool) =
  if self.boldInternal != value:
    self.boldInternal = value
    discard self.toCacheKey()

proc italic*(self: TextFormat): bool = self.italicInternal
proc `italic=`*(self: TextFormat, value: bool) =
  if self.italicInternal != value:
    self.italicInternal = value
    discard self.toCacheKey()

proc newTextFormat*(font: string = "", size: int = 12, color: int = 0, bold: bool = false, italic: bool = false, underline: bool = false, url: string = "", target: string = "", align: TextFormatAlign = TextFormatAlign.LEFT, leftMargin: int = 0, rightMargin: int = 0, indent: int = 0, leading: int = 0): TextFormat =
  let self = TextFormat(
    fontInternal: font,
    sizeInternal: size,
    color: color,
    boldInternal: bold,
    italicInternal: italic,
    underline: underline,
    url: url,
    target: target,
    align: align,
    leftMargin: leftMargin,
    rightMargin: rightMargin,
    indent: indent,
    leading: leading,
    tabStops: @[]
  )
  discard self.toCacheKey()
  return self

proc clone*(self: TextFormat): TextFormat =
  let newFormat = newTextFormat(self.fontInternal, self.sizeInternal, self.color, self.boldInternal, self.italicInternal, self.underline, self.url, self.target)
  newFormat.align = self.align
  newFormat.leftMargin = self.leftMargin
  newFormat.rightMargin = self.rightMargin
  newFormat.indent = self.indent
  newFormat.leading = self.leading
  newFormat.blockIndent = self.blockIndent
  newFormat.bullet = self.bullet
  newFormat.kerning = self.kerning
  newFormat.letterSpacing = self.letterSpacing
  newFormat.tabStops = self.tabStops
  newFormat.strikethrough = self.strikethrough
  newFormat.cacheKey = self.cacheKey
  return newFormat

File: kfl/src/main/kotlin/openfl/text/TextFormat.kt
package openfl.text

class TextFormat(
    var font: String? = null,
    var size: Int? = null,
    var color: Int? = null,
    var bold: Boolean? = null,
    var italic: Boolean? = null,
    var underline: Boolean? = null,
    var url: String? = null,
    var target: String? = null,
    var align: TextFormatAlign? = null,
    var leftMargin: Int? = null,
    var rightMargin: Int? = null,
    var indent: Int? = null,
    var leading: Int? = null
) {
    var blockIndent: Int? = null
    var bullet: Boolean? = null
    var kerning: Boolean? = null
    var letterSpacing: Double? = null
    var tabStops: Array<Int>? = null
    var strikethrough: Boolean? = null

    fun clone(): TextFormat {
        val newFormat = TextFormat(font, size, color, bold, italic, underline, url, target, align, leftMargin, rightMargin, indent, leading)
        newFormat.blockIndent = blockIndent
        newFormat.bullet = bullet
        newFormat.kerning = kerning
        newFormat.letterSpacing = letterSpacing
        newFormat.tabStops = tabStops
        newFormat.strikethrough = strikethrough
        return newFormat
    }

    internal fun __merge(format: TextFormat) {
        if (format.font != null) font = format.font
        if (format.size != null) size = format.size
        if (format.color != null) color = format.color
        if (format.bold != null) bold = format.bold
        if (format.italic != null) italic = format.italic
        if (format.underline != null) underline = format.underline
        if (format.url != null) url = format.url
        if (format.target != null) target = format.target
        if (format.align != null) align = format.align
        if (format.leftMargin != null) leftMargin = format.leftMargin
        if (format.rightMargin != null) rightMargin = format.rightMargin
        if (format.indent != null) indent = format.indent
        if (format.leading != null) leading = format.leading
        if (format.blockIndent != null) blockIndent = format.blockIndent
        if (format.bullet != null) bullet = format.bullet
        if (format.kerning != null) kerning = format.kerning
        if (format.letterSpacing != null) letterSpacing = format.letterSpacing
        if (format.tabStops != null) tabStops = format.tabStops
        if (format.strikethrough != null) strikethrough = format.strikethrough
    }
}
